### ISSUE TITLE
Extracting colored point clouds

### ISSUE THREAD

**Issue by orweiser-code** [NVlabs/GaussianSTORM#3]
Hey great work, first of all!
i was wondering about extracting the 3D pointcloud, per timestamp.
1. i have extracted the gaussian means and colors (and map with colors = colors * 0.5 + 0.5 to [0,1]) from gs_params, but the colors are kind of funky - a pinkish grey that is not correct for sure, and anyway one should deal with the opacity of the gaussians.

![Image](https://github.com/user-attachments/assets/b574f8a7-ef8d-4009-8ae5-583fe371edce)

2. i think the better way would be to backproject the rendered depth, and use the colors rendered to the images. do you have a code snippet to do that using the data structures in your code?

thanks

**Comment by kevinchiu19** [comment]
For colors, you should use the affine transformation estimated by the network, refer here:
https://github.com/NVlabs/GaussianSTORM/blob/fcd2561de352f756bb59a66655d62b3af4c2c39c/storm/models/storm.py#L624

**Comment by orweiser-code** [comment]
Thanks for the quick response. The colors look good now!
what do you think about point 2? or maybe another way to query the amodal gaussians for 3d structure at some time point t?

**Comment by rain72510** [comment]
@orweiser-code 
Hello, I'm trying to visualize the gaussians too. May I ask that how you do the affine transformation?
Currently what I do is
```python
C0 = 0.28209479177387814      # 1 /(2sqrt pi)
...
if gs_params.get("use_affine_token", True) and "affine" in gs_params:
        affine = gs_params["affine"].mean(1).cpu()
        rgb_lin = (colors + 1) * 0.5
        ones    = torch.ones_like(rgb_lin[..., :1])
        rgb_aug = torch.cat([rgb_lin, ones], dim=-1)
        colors  = (rgb_aug @ affine[0].T)
    else:
        colors  = (colors + 1) * 0.5

rgb01   = colors.clamp(0, 1).float()
f_dc_np = ((rgb01 - 0.5) / C0).numpy()
...
vert["f_dc_0"], vert["f_dc_1"], vert["f_dc_2"]    = f_dc_np.T
```
But this makes my Gaussians looks all green; I use online website [supersplat](https://superspl.at/editor) to visualize

![Image](https://github.com/user-attachments/assets/c3d5d1cf-45b0-4e90-812c-95053b3f4aa0)

Thanks!

**Comment by finninmunich** [comment]
> [@orweiser-code](https://github.com/orweiser-code) Hello, I'm trying to visualize the gaussians too. May I ask that how you do the affine transformation? Currently what I do is
> 
> C0 = 0.28209479177387814      # 1 /(2sqrt pi)
> ...
> if gs_params.get("use_affine_token", True) and "affine" in gs_params:
>         affine = gs_params["affine"].mean(1).cpu()
>         rgb_lin = (colors + 1) * 0.5
>         ones    = torch.ones_like(rgb_lin[..., :1])
>         rgb_aug = torch.cat([rgb_lin, ones], dim=-1)
>         colors  = (rgb_aug @ affine[0].T)
>     else:
>         colors  = (colors + 1) * 0.5
> 
> rgb01   = colors.clamp(0, 1).float()
> f_dc_np = ((rgb01 - 0.5) / C0).numpy()
> ...
> vert["f_dc_0"], vert["f_dc_1"], vert["f_dc_2"]    = f_dc_np.T
> But this makes my Gaussians looks all green; I use online website [supersplat](https://superspl.at/editor) to visualize
> 
> ![Image](https://github.com/user-attachments/assets/c3d5d1cf-45b0-4e90-812c-95053b3f4aa0)
> 
> Thanks!

Hi，can you share the script for generating supersplat-supported ply file? I am a beginner of 3DGS and I can't produce the correct ply file for supersplat. This will help me a lot, 100 thanks for you!!

**Comment by finninmunich** [comment]
> > [@orweiser-code](https://github.com/orweiser-code) Hello, I'm trying to visualize the gaussians too. May I ask that how you do the affine transformation? Currently what I do is
> > C0 = 0.28209479177387814      # 1 /(2sqrt pi)
> > ...
> > if gs_params.get("use_affine_token", True) and "affine" in gs_params:
> > affine = gs_params["affine"].mean(1).cpu()
> > rgb_lin = (colors + 1) * 0.5
> > ones    = torch.ones_like(rgb_lin[..., :1])
> > rgb_aug = torch.cat([rgb_lin, ones], dim=-1)
> > colors  = (rgb_aug @ affine[0].T)
> > else:
> > colors  = (colors + 1) * 0.5
> > rgb01   = colors.clamp(0, 1).float()
> > f_dc_np = ((rgb01 - 0.5) / C0).numpy()
> > ...
> > vert["f_dc_0"], vert["f_dc_1"], vert["f_dc_2"]    = f_dc_np.T
> > But this makes my Gaussians looks all green; I use online website [supersplat](https://superspl.at/editor) to visualize
> > ![Image](https://github.com/user-attachments/assets/c3d5d1cf-45b0-4e90-812c-95053b3f4aa0)
> > Thanks!
> 
> Hi，can you share the script for generating supersplat-supported ply file? I am a beginner of 3DGS and I can't produce the correct ply file for supersplat. This will help me a lot, 100 thanks for you!!

Can use gsplat.export_splats for supersplat-supported ply format file generation

<img width="1113" height="611" alt="Image" src="https://github.com/user-attachments/assets/424baf18-9306-453b-bfbc-ce51bb6e6a18" />

**Comment by Chromium0516** [comment]
> > > [@orweiser-code](https://github.com/orweiser-code) Hello, I'm trying to visualize the gaussians too. May I ask that how you do the affine transformation? Currently what I do is
> > > C0 = 0.28209479177387814      # 1 /(2sqrt pi)
> > > ...
> > > if gs_params.get("use_affine_token", True) and "affine" in gs_params:
> > > affine = gs_params["affine"].mean(1).cpu()
> > > rgb_lin = (colors + 1) * 0.5
> > > ones    = torch.ones_like(rgb_lin[..., :1])
> > > rgb_aug = torch.cat([rgb_lin, ones], dim=-1)
> > > colors  = (rgb_aug @ affine[0].T)
> > > else:
> > > colors  = (colors + 1) * 0.5
> > > rgb01   = colors.clamp(0, 1).float()
> > > f_dc_np = ((rgb01 - 0.5) / C0).numpy()
> > > ...
> > > vert["f_dc_0"], vert["f_dc_1"], vert["f_dc_2"]    = f_dc_np.T
> > > But this makes my Gaussians looks all green; I use online website [supersplat](https://superspl.at/editor) to visualize
> > > ![Image](https://github.com/user-attachments/assets/c3d5d1cf-45b0-4e90-812c-95053b3f4aa0)
> > > Thanks!
> > 
> > 
> > Hi，can you share the script for generating supersplat-supported ply file? I am a beginner of 3DGS and I can't produce the correct ply file for supersplat. This will help me a lot, 100 thanks for you!!
> 
> Can use gsplat.export_splats for supersplat-supported ply format file generation
> 
> <img alt="Image" width="1113" height="611" src="https://private-user-images.githubusercontent.com/75730191/526379147-424baf18-9306-453b-bfbc-ce51bb6e6a18.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjYxNTY5NDcsIm5iZiI6MTc2NjE1NjY0NywicGF0aCI6Ii83NTczMDE5MS81MjYzNzkxNDctNDI0YmFmMTgtOTMwNi00NTNiLWJmYmMtY2U1MWJiNmU2YTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE5VDE1MDQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyM2Q3ODhhYjM0MGE2ZGZjMDBkYThlNjM0ZjBhODMxZDAwZGY5ZTg2MmI1OGNmYWFlOWFiZDU1OTkzNzZmZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M3Lb8ICO3jHimTlLkmVigSXYFktIg-UByhS04erehUo">

How do you do that？Can you share the script？Thank you so much！！！

**Comment by finninmunich** [comment]
> > > > [@orweiser-code](https://github.com/orweiser-code) Hello, I'm trying to visualize the gaussians too. May I ask that how you do the affine transformation? Currently what I do is
> > > > C0 = 0.28209479177387814      # 1 /(2sqrt pi)
> > > > ...
> > > > if gs_params.get("use_affine_token", True) and "affine" in gs_params:
> > > > affine = gs_params["affine"].mean(1).cpu()
> > > > rgb_lin = (colors + 1) * 0.5
> > > > ones    = torch.ones_like(rgb_lin[..., :1])
> > > > rgb_aug = torch.cat([rgb_lin, ones], dim=-1)
> > > > colors  = (rgb_aug @ affine[0].T)
> > > > else:
> > > > colors  = (colors + 1) * 0.5
> > > > rgb01   = colors.clamp(0, 1).float()
> > > > f_dc_np = ((rgb01 - 0.5) / C0).numpy()
> > > > ...
> > > > vert["f_dc_0"], vert["f_dc_1"], vert["f_dc_2"]    = f_dc_np.T
> > > > But this makes my Gaussians looks all green; I use online website [supersplat](https://superspl.at/editor) to visualize
> > > > ![Image](https://github.com/user-attachments/assets/c3d5d1cf-45b0-4e90-812c-95053b3f4aa0)
> > > > Thanks!
> > > 
> > > 
> > > Hi，can you share the script for generating supersplat-supported ply file? I am a beginner of 3DGS and I can't produce the correct ply file for supersplat. This will help me a lot, 100 thanks for you!!
> > 
> > 
> > Can use gsplat.export_splats for supersplat-supported ply format file generation
> > <img alt="Image" width="1113" height="611" src="https://private-user-images.githubusercontent.com/75730191/526379147-424baf18-9306-453b-bfbc-ce51bb6e6a18.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjYxNTY5NDcsIm5iZiI6MTc2NjE1NjY0NywicGF0aCI6Ii83NTczMDE5MS81MjYzNzkxNDctNDI0YmFmMTgtOTMwNi00NTNiLWJmYmMtY2U1MWJiNmU2YTE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjE5VDE1MDQwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyM2Q3ODhhYjM0MGE2ZGZjMDBkYThlNjM0ZjBhODMxZDAwZGY5ZTg2MmI1OGNmYWFlOWFiZDU1OTkzNzZmZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.M3Lb8ICO3jHimTlLkmVigSXYFktIg-UByhS04erehUo">
> 
> How do you do that？Can you share the script？Thank you so much！！！

Check the rendering function in this repo ( gsplat.rasterization) , there is a function called export_splats which you can also import from gsplat. Their inputs are familiar.
For the color, remember to use affine transformation; For scale, remember to use log scale
